[Telemetry] Add event metrics for HotPOKT

derrandz opened the pull request:
## Description


This PR adds event-metrics to the HotPOKT consensus implementation using the telemetry module.

## Changes
Please mark the options that are relevant.

- [x] Added instrumentation to record the reception of a hotpokt message as an event for each block, this event has the following attributes:
     - namespace `consensus`
     - event_name `hotpokt_messages`
     - labels: [height, step, validator_role]

- [x] Added logql query to process the recorded events and fed the results to a numerical series graph

## How Has This Been Tested?
__

1. Spin up the stack
```zsh
$ make compose_and_watch
```

2. Make sure you spin the client up, and that kick the blockchain off:
```zsh
$ make client_start && make client_connect
```

3. Choose:
```
✔ TogglePacemakerMode
✔ TriggerNextView
```

4. Open Grafana at [https://localhost:3000](https://localhost:3000)
5. Open the `HotPokt Telemetry Graphs` Dashboard

6. Verify the results on the dashboard

### Additional items for review

The query that was used to generate the graph is the following:
```logql
```

So I need you to also revise the logic in this query.


## Owners
@derrandz

Olshansk committed:
Pocket Network Consensus Module Specification - v1.0.1

The first version of the HotPocket specification.

Olshansk committed:
Friday night WIP for RainTree happy path no-tests on top of Pre2P - almost working

Olshansk committed:
LocalNet is running but propagation isn't fully functional yet - only reaching 50% of nodes. Need to add debug logs and fix the happy path

Olshansk committed:
Almost working but not progressing on every height

Olshansk committed:
Validator configuration improvements

Olshansk committed:
WIP

Olshansk committed:
Add prototype back

Olshansk committed:
Happy path ACTUALLY working now

Olshansk committed:
Added TransportLayerConn prior to writing tests

Olshansk committed:
Replaced Accept w/ TCP in TransportLayerConn abstraction

Olshansk committed:
Pipes are a difficult abstraction - going to try channels

Olshansk committed:
GOT A TEST WORKING - time to sleep

Olshansk committed:
Enforce the correct number of reads & writes

Olshansk committed:
Fix race condition

Olshansk committed:
Add proper demote logic and improve tests

Olshansk committed:
Small & Large RainTree test working consistenly

Olshansk committed:
Cleanup - replaced 'pipeConn' with 'emptyConn'

Olshansk committed:
Rename RawData to NetworkData

Olshansk committed:
Cleanup - self review of raintree/network code

Olshansk committed:
Cleanup - self review of raintree test

Olshansk committed:
Cleanup - removed and ignoring network_mock.go from git history

Olshansk committed:
Fixed genesis configs tests

Olshansk committed:
Fix raintree_test expected num reads & writes

Commit:
implemented the cleanup and redundancy layer happy paths

okdas committed:
add victoria metrics and configure docker SD

okdas committed:
some onboarding changes + relabel configs

derrandz committed:
module: add telemetry module with prometheus metrics sdk

derrandz committed:
feat: track pre2p metrics

okdas committed:
listen on all interfaces

derrandz committed:
fix: expose http metrics endpoint on external net interface (docker)

derrandz committed:
add more metrics to track and allow switching telemetry on and off to avoid issues with the client

okdas committed:
add more containers

okdas committed:
add loki

derrandz committed:
add missing quotes to expose ports in docker-compose

derrandz committed:
chore: move to laptop

derrandz committed:
feat:
  - add NOOP telemetry module to enable telemetry-less nodes/clients
  - reduce metric list to two (blockheight + total_received_broadcast_msgs)
  - add exported grafana panel in json format

okdas committed:
fixup configs & cleanup

okdas committed:
Fix scraping config regex

derrandz committed:
reduce validators to 4 to let the blockchain go forward (bug?)

derrandz committed:
change machine, revert this commit

derrandz committed:
feat: use logging to track events, add grafana dashbaord to aggregate log events and generate graphs

derrandz committed:
chore: remove non-related changes

derrandz committed:
chore: cleanup

derrandz committed:
chore: cleanup

derrandz committed:
feat: add timeseries metrics

Olshansk commented:
@derrandz
<img width="322" alt="Screen Shot 2022-06-21 at 8 38 03 PM" src="https://user-images.githubusercontent.com/1892194/174938359-3e5a7069-f15b-4bec-be10-f60fe8842470.png">

Pre2P and HotPOKT messages? Is this just a typo or intended?

derrandz commented:
> @derrandz
> <img width="322" alt="Screen Shot 2022-06-21 at 8 38 03 PM" src="https://user-images.githubusercontent.com/1892194/174938359-3e5a7069-f15b-4bec-be10-f60fe8842470.png">
> 
> Pre2P and HotPOKT messages? Is this just a typo or intended? 

It was. Thanks for catching that.

derrandz committed:
add back removed files

derrandz committed:
implement requested changes

derrandz committed:
fix any keyword syntax and upgrade docker image to 1.18

okdas committed:
rm build/granafa/provisioning/datasources/datasource.yaml

okdas committed:
This config is actually also not in use.

derrandz committed:
add requested changes

derrandz committed:
correct rebase slip

derrandz committed:
add back missing changes

derrandz committed:
add back missing changes

Olshansk committed:
Remove redundancy in 'make refresh' target

Olshansk committed:
Updated the documentation

Olshansk reviewed (changes requested):
@derrandz Before I review this PR, please see the following:

1. Comments from https://github.com/pokt-network/pocket/pull/96#pullrequestreview-1026680551 are pertinent here too. Labels, whitespace, constants, etc...

2. Since this PR 97 is dependent on PR 96, you should use that as a base so it is easier for the reviewer to understand what's being submitted / reviewed ONLY here.

derrandz committed:
use the tx mempool for p2p msg de-dupping

derrandz committed:
add a few DISCUSS comments to keep track of our tech debt

derrandz committed:
feat: split up the module interface to allow subcomponents to use bus functionality through the sub interface: IntegratableModule

derrandz committed:
feat: factorize all metric name definitions into one file

Olshansk commented:
@derrandz After we merge in the `RainTree` PR, could we do apply the same changes here?

derrandz committed:
move metric definitions into a separate file

derrandz commented:
@Olshansk Sure thing, Just rebased and added the similar changes to this branch. Off to you.

derrandz committed:
[Telemetry] Add event metrics for RainTree (#96)

This PR adds event-metrics through to Pre2P RainTree implementation using the Telemetry module.

- [x] Added instrumentation to record the reception of a RainTree message as an event for each block, this event has the following attributes:
     - namespace `pre2p`
     - event_name `raintree_messages`
     - labels: [height]

- [x] Added instrumentation to record the reception of redundant RainTree messages as an event for each new block, the event has the following attributes:
     - namespace `pre2p`
     - event_name `broadcast_message_redundancy_per_block`
     - labels: [height, message_hash]

- [x] Added `logql` query to process the recorded events and fed the results to a numerical series graph

__

1. Spin up the stack
```zsh
$ make compose_and_watch
```

2. Make sure you spin the client up, and that kick the blockchain off:
```zsh
$ make client_start && make client_connect
```

3. Choose:
```
✔ TogglePacemakerMode
✔ TriggerNextView
```

4. Open Grafana at [http://localhost:3000](http://localhost:3000)
5. Open the `RainTree Telemetry` graph

6. Verify the results on the dashboard

The query that was used to generate the graph is the following:

1. Total # RainTree Msgs / Block
```sql
sum by (blockheight) (
   count_over_time({host="docker-desktop"}
   |= "[EVENT] consensus"
   | pattern `<datetime> <_> <time> <type> <_> <metric_name> <_> <blockheight> <step> <validator_role>`[6s]
   )
),
```

2. Avg # Raintree Msgs / Block

```sql
avg by (blockheight, instance) (
  sum by (blockheight) (
    count_over_time({host="docker-desktop"}
    |= "[EVENT] p2p raintree_messages"
    | pattern `<datetime> <_> <time> <type> <namespace> <metric> <_> <blockheight>`[6s])
  )
)
```

3. Avg. Redundancy / Block
```sql
avg by (blockheight, instance) (
  sum by (blockheight) (
    count_over_time({host="docker-desktop"}
    |= "[EVENT] p2p broadcast_message_redundancy_per_block"
    | pattern `<datetime> <_> <time> <type> <_> <redundancy> <_> <hashstring> <_> <blockheight>`[6s])
   )
)
```

Owners: @derrandz
Co-owners: @okdas @Olshansk

Olshansk commented on consensus/hotstuff_replica.go:
```diff
@@ -136,6 +192,12 @@ func (handler *HotstuffReplicaMessageHandler) HandleDecideMessage(m *consensusMo
 	}
 
 	m.paceMaker.NewHeight()
+	m.GetBus().
```
If you move this into `NewHeight`, then we don't need to call the same counter increment in both the leader & replica code

Olshansk commented on consensus/telemetry/metrics.go:
```diff
@@ -0,0 +1,30 @@
+package consensus_telemetry
+
+const (
+	// How to define metrics in this file?
```
Can this be moved into a shared telemetry README (since it'll be the same for every module) and reference that here?

Please apply this to any other places too.

Olshansk commented on p2p/pre2p/module.go:
```diff
@@ -9,6 +9,7 @@ import (
 
 	"github.com/pokt-network/pocket/p2p/pre2p/raintree"
 	"github.com/pokt-network/pocket/p2p/pre2p/stdnetwork"
+	pre2pTelemetry "github.com/pokt-network/pocket/p2p/pre2p/telemetry"
```
Make sure to rebase the integration branch with main later. Pre2P is no more

Olshansk commented on p2p/pre2p/module.go:
```diff
@@ -77,6 +78,16 @@ func (m *p2pModule) GetBus() modules.Bus {
+		GetTelemetryModule().
+		GetTimeSeriesAgent().
+		CounterRegister(
+			pre2pTelemetry.P2P_NODE_STARTED_TIMESERIES_METRIC_NAME,
```
The title of the PR says "HotPokt", the description says "raintree" and the code includes both. Which is it?

Olshansk commented on p2p/pre2p/telemetry/metrics.go:
```diff
@@ -0,0 +1,20 @@
+	// Must define metrics using following template: MetricName_MetricType_Attribute
+	// e.g: type= timeseries, name=nodes_alive_counter, attribute=name
+	// will result in: NODES_COUNT_COUNTER_TIME_SERIES_NAME="nodes_alive_counter"
+	// you can also add: NODES_COUNT_COUNTER_TIME_SERIES_DESCRIPTION="for your description"
```
Copy-pasta other comment: Only have this comment in one place (i.e. the telemetry readme)

Olshansk commented on shared/modules/module.go:
```diff
@@ -2,9 +2,16 @@ package modules
 
 // TODO(olshansky): Show an example of `TypicalUsage`
 type Module interface {
-	Start() error
```
Did you rebase? I thought this was added in the other PR.

Olshansk reviewed (changes requested):
The title says "HotPOKT" but the description includes "pre2p raintree". Please review and update appropriately

Olshansk commented:
@derrandz Any chance you could resolve conflicts and resubmit for review? 

Here is the current delta:
![Screen Shot 2022-08-01 at 1 42 35 PM](https://user-images.githubusercontent.com/1892194/182243326-9c23a1b3-a292-489e-92b6-1feb73489d6e.png)

derrandz committed:
cleanup rebase mistakes and leftovers

derrandz committed:
remove documentation comment in favor of README file at shared/telemetry/README.md

derrandz committed:
DRY blockchain height metric tracking

derrandz committed:
fix dashboard issues caused by rebase

Olshansk reviewed (commented):
LGTM. :shipit:
